pkg/endpoint/bpf.go

```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 		e.PolicyDebug(logrus.Fields{"dumpedDiffs": diffs}, "syncPolicyMapWithDump")
 	}
 
+	if err == nil {
+		e.policyMapLastFullReconciliation = time.Now()
+	}
+
 	return err
 }
 
-func (e *Endpoint) startSyncPolicyMapController() {
-	// Skip the controller if the endpoint has no policy map
-	if !e.HasBPFPolicyMap() {
-		return
+func (e *Endpoint) syncPolicyMapControllerName() string {
+	return fmt.Sprintf("sync-policymap-%d", e.ID)
+}
+
+// updateSyncPolicyMapController creates or triggers a controller to  reconcile
+// pending policy map changes. Every PolicyMapFullReconciliationInterval the
+// controller also does full reconciliation by dumping the policy map.
+func (e *Endpoint) updateSyncPolicyMapController() channels.DoneChan {
+	// Skip the controller if the endpoint has no policy map or is
+	// being disconnected.
+	if !e.HasBPFPolicyMap() || e.IsDisconnecting() {
+		return channels.ClosedDoneChan
 	}
 
-	ctrlName := fmt.Sprintf("sync-policymap-%d", e.ID)
-	e.controllers.CreateController(ctrlName,
+	return e.controllers.UpdateController(
+		e.syncPolicyMapControllerName(),
 		controller.ControllerParams{
 			Group:          syncPolicymapControllerGroup,
 			HealthReporter: e.GetReporter("policymap-sync"),
 			DoFunc: func(ctx context.Context) error {
-				// Failure to lock is not an error, it means
-				// that the endpoint was disconnected and we
-				// should exit gracefully.
-				if err := e.lockAlive(); err != nil {
+				e.unconditionalLock()
+				defer e.unlock()
+				if e.IsDisconnecting() {
 					return controller.NewExitReason("Endpoint disappeared")
 				}
-				defer e.unlock()
+
 				if e.desiredPolicy != e.realizedPolicy {
 					// Currently in the middle of a regeneration; do not execute
 					// at this time.
 					return nil
 				}
-				return e.syncPolicyMapWithDump()
+
+				err := e.syncPolicyMapWithDump(ctx)
+				if err != nil {
+					return err
+				}
+
+				// Apply policy changes to L7 proxy
+				proxyWaitGroup := completion.NewWaitGroup(ctx)
+
+				// Ignoring the revertFunc; keep all successful changes even if some fail.
+				if err, _ = e.updateNetworkPolicy(proxyWaitGroup); err != nil {
+					e.getLogger().WithError(err).Warn("failed to publish policy update to L7 proxy")
+				}
+				proxyWaitGroup.Wait()
```

Comment:
Why did this get moved into the controller DoFunc? It wasn't there previously.

Reply:
Right, this hits at a very subtle concurrency change.

Previously, this was divided in to three phases:

1. Queue policy notifications via `selectorcache.UpdateFQDNSelector() / .UpdateIdentities()`
2. Wait for notifications to be complete. Then, go `ep.UpdatePolicyMaps()` for all endpoints. Queue updates for Envoy
3. Wait for Envoy to finish updating.

We do lose a bit of parallelism, as we can make progress with the the next (2) as soon as it is done, rather than waiting for Envoy to ack any updates. I don't have a good intuition for whether or not this is relevant. My instinct is no; updating policy maps is not particularly complicated, so we probably don't gain much by making progress on the next iteration while waiting for Envoy. However, it would be good to reason about this.

Reply:
> Why did this get moved into the controller DoFunc? It wasn't there previously.

Refactored to use another controller to apply the policy map pending changes.

> My instinct is no; updating policy maps is not particularly complicated, so we probably don't gain much by making progress on the next iteration while waiting for Envoy. However, it would be good to reason about this.

My hunch is that your intuition is correct.
Moreover, using a single controller to apply policy map pending changes **and** push them to the L7 proxy makes it impossible to get a notification for the completion of (2) before executing (3).
For now, I kept this approach with the dedicated controller too, but if you have better ideas I'm happy to hear them.